README.md
# redshift-query

[![image](https://img.shields.io/pypi/v/redshift-query.svg)](https://pypi.python.org/pypi/redshift-query)
[![image](https://img.shields.io/travis/helecloud/redshift-query.svg)](https://travis-ci.com/helecloud/redshift-query)
[![image](https://readthedocs.org/projects/redshift-query/badge/?version=latest)](https://redshift-query.readthedocs.io/en/latest/?badge=latest)


## Introduction

This is a very simple library that gets credentials of a cluster via
[redshift.GetClusterCredentials](https://docs.aws.amazon.com/redshift/latest/APIReference/API_GetClusterCredentials.html)
API call and then makes a connection to the cluster and runs the provided SQL statements, once done it will close the connection and return the results.

This is useful for when you want to run queries in CLIs or based on events for example on AWS Lambdas, or on a regular
basis on AWS Glue Python Shell jobs.

## Usage

While redshift_query could be used as a library it or it's own standalone script.

It requires the following parameters supplied either as environmental variables or keys in the parameters dict passed to the function:

- DB_NAME: The database name. You can also provide this via db_name parameter to the function.
- DB_USER: The user to request credentials from. You can optionally provide this via db_user parameter to the function.
- CLUSTER_HOST: The host name of the redshift cluster. You can optionally provide this via cluster_host parameter to the function.
- CLUSTER_ID: The id of the cluster. You can optionally provide this via cluster_id parameter to the function.
- SQL_STATEMENTS: The SQL statements to run. You can optionally provide this via sql_statements parameter to the function.
This parameter is going to be formatted(via [string.format](https://docs.python.org/3/library/string.html)) with the event object that is passed to it.
This way you can have the SQL statement to be based of the event that's passed to your function, which is useful
for when you have a Lambda and it is called by an event source(S3 for example).
- REDSHIFT_QUERY_LOG_LEVEL: By default set to ERROR, which logs nothing. Normally errors are not logged and bubbled up instead so they crash the script.
If set to INFO, it will log the result of queries and if set to DEBUG it will log every thing that happens which is good for debugging why it is stuck.

### Deploying via AWS SAM & Lambda

Here's an example that copies to a table when new manifests in S3 are written:

```yaml
AWSTemplateFormatVersion: '2010-09-09'
Transform: AWS::Serverless-2016-10-31
Resources:
  Function:
    Type: AWS::Serverless::Function
    Properties:
      CodeUri: ./redshift_query/redshift_query.py
      Handler: redshift_query.query
      Runtime: python3.8
      VpcConfig:
        SecurityGroupIds: 'sg-12312312'
        SubnetIds: 'sub-123123'
      Policies:
        - Version: '2012-10-17'
          Statement:
            - Effect: "Allow"
              Action: "redshift:GetClusterCredentials"
              Resource: "arn:aws:redshift:eu-west-1:123123123:dbuser:test/master" # https://docs.aws.amazon.com/redshift/latest/mgmt/generating-iam-credentials-role-permissions.html
      Events:
        S3:
          Type: Schedule
          Properties:
            Bucket: mybucket
            Events: s3:ObjectCreated:*
            Filter:
              S3Key:
                Rules:
                  - Name: suffix
                    Value: .manifest

      Environment:
        Variables:
          CLUSTER_ID: 'test'
          CLUSTER_HOST: 'test.j242tj1qkjuz.eu-west-1.redshift.amazonaws.com'
          DB_NAME: 'test'
          DB_USER: 'master'
          SQL_STATEMENTS: |-
            copy customer
              from {Records[0]['s3']['object']['key']}
            iam_role 'arn:aws:iam::0123456789012:role/MyRedshiftRole'
            manifest
          LOG_LEVEL: DEBUG
```

### Usage as a Library

To use the library directly you can simply provide the parameters in snake_case format:
```python
import redshift_query

results = redshift_query.query({
    'db_name': 'test',
    'db_user': 'master',
    'cluster_host': 'test.j242tj1qkjuz.eu-west-1.redshift.amazonaws.com',
    'cluster_id': 'test',
    'sql_statements': '''
        copy customer
            from 's3://mybucket/cust.manifest'
            iam_role 'arn:aws:iam::0123456789012:role/MyRedshiftRole'
            manifest;
        select * from customer;
     '''
})

[copy_results, select_results] = results

print('select results', select_results)
```

If you use redshift_query.query multiple times in your code you can use redshift_query.set_config to set the static configuration once:

```python
import redshift_query

redshift_query.set_config({
    'db_name': 'test',
    'db_user': 'master',
    'cluster_host': 'test.j242tj1qkjuz.eu-west-1.redshift.amazonaws.com',
    'cluster_id': 'test'
})

redshift_query.query({
    'sql_statements': '''
        copy customer
            from 's3://mybucket/cust.manifest'
            iam_role 'arn:aws:iam::0123456789012:role/MyRedshiftRole'
            manifest;
     '''
})
```

Optionally you can also provide boto_session, if you want to assume role or provide your own credentials:

```python
import redshift_query
import boto3

redshift_query.query({
    'boto_session': boto3.session.Session(profile_name='production'),
    'sql_statements': 'select 1;'
})
```

### As a CLI

```
$ DB_USER=neshat DB_NAME=test CLUSTER_ID=test CLUSTER_HOST=test.3j232jjqji21.eu-west-1.redshift.amazonaws.com SQL_STATEMENTS='select 1;' REDSHIFT_QUERY_LOG_LEVEL=DEBUG redshift_query
DEBUG:redshift_query:Passed Event: {}
DEBUG:redshift_query:Passed this_config with added defaults: {'db_name': 'test', 'db_user': 'neshat', 'cluster_host': 'test.3j232jjqji21.eu-west-1.redshift.amazonaws.com', 'cluster_id': 'test', 'sql_statements': 'select 1;', 'boto_session': None}
DEBUG:redshift_query:Received Credentials
DEBUG:redshift_query:Connected
DEBUG:redshift_query:Running select 1
INFO:redshift_query:Statement "?column?
--------
       1
(1 row)" result:
```

Credits
-------

This package was created with
[Cookiecutter](https://github.com/audreyr/cookiecutter) and the
[audreyr/cookiecutter-pypackage](https://github.com/audreyr/cookiecutter-pypackage)
project template.


HISTORY.rst
=======
History
=======

0.0.3 (2020-07-13)
------------------

* First release on PyPI.


tests/__init__.py
"""Unit test package for redshift_query."""


tests/test_redshift_query.py
#!/usr/bin/env python

"""Tests for `redshift_query` package."""

from unittest.mock import patch, call
import unittest
import boto3
import botocore.session
import pg
from botocore.stub import Stubber
from redshift_query import redshift_query
import logging

logging.basicConfig()


class MockBoto(boto3.session.Session):
    def __init__(self, expected_params: dict):
        super().__init__()
        self.expected_params = expected_params

    def client(self, name: str, **kwargs):
        redshift = botocore.session.get_session().create_client('redshift', region_name='us-east-1')
        stubber = Stubber(redshift)

        stubber.add_response('get_cluster_credentials', {
            'DbUser': 'user',
            'DbPassword': 'password'
        }, self.expected_params)

        stubber.activate()

        return redshift


class TestRedshiftQuery(unittest.TestCase):
    """Tests for `redshift_query` package."""

    def setUp(self):
        """Set up test fixtures, if any."""

    def tearDown(self):
        """Tear down test fixtures, if any."""

    def test_query(self):
        with patch.object(pg, 'DB') as mock_method:
            mock_method().query.return_value = 'test'
            results = redshift_query.query({
                'db_name': 'name',
                'db_user': 'user',
                'cluster_host': 'host',
                'cluster_id': 'id',
                'sql_statements': 'select 1;',
                "boto_session": MockBoto(expected_params={
                    'DbUser': 'user',
                    'DbName': 'name',
                    'ClusterIdentifier': 'id',
                    'AutoCreate': False
                })
            })

        mock_method.assert_has_calls([
            call("host=host dbname=name user=user password=password port=5439 keepalives=1 "
                 "keepalives_idle=200 keepalives_interval=200 keepalives_count=6 "
                 "connect_timeout=10"),
            call().query('select 1'),
            call().close()
        ], any_order=True)

        self.assertEqual(results, ['test'])


redshift_query/redshift_query.py
# MIT License
#
# Copyright (c) 2020 Helecloud
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import pg
import boto3
import os
import logging

config = {
    'db_name': os.getenv('DB_NAME') or None,
    'db_user': os.getenv('DB_USER'),
    'cluster_host': os.getenv('CLUSTER_HOST'),
    'cluster_id': os.getenv('CLUSTER_ID'),
    'sql_statements': os.getenv('SQL_STATEMENTS'),
    'boto_session': None
}

logger = logging.getLogger('redshift_query')
logger.setLevel(os.getenv('REDSHIFT_QUERY_LOG_LEVEL', 'ERROR'))


def set_config(new_config):
    global config
    config = {**config, **new_config}


def query(event, context=None):
    logger.debug('Passed Event: %s', event)

    this_config = {**config, **event}

    session = this_config['boto_session'] or boto3.session.Session()

    if not isinstance(session, boto3.session.Session):
        raise RuntimeError('Not a valid boto3 session. Please check your configuration')

    client = session.client('redshift')

    logger.debug('Passed this_config with added defaults: %s', this_config)

    credentials = client.get_cluster_credentials(
        DbUser=this_config['db_user'],
        DbName=this_config['db_name'],
        ClusterIdentifier=this_config['cluster_id'],
        AutoCreate=False,
    )

    logger.debug('Received Credentials')

    db = pg.DB(f"host={this_config['cluster_host']} "
               f"dbname={this_config['db_name']} "
               f"user={credentials['DbUser']} "
               f"password={credentials['DbPassword']} "
               'port=5439 '
               'keepalives=1 keepalives_idle=200 keepalives_interval=200 keepalives_count=6 '
               'connect_timeout=10')

    logger.debug('Connected')

    statements = this_config['sql_statements'].format(**event).strip().split(';')

    def run(statement):
        logger.debug('Running %s', statement)
        result = db.query(statement)
        logger.info('Statement "%s" result:', result)
        return result

    results = [
        run(statement)
        for statement in statements
        if statement
    ]

    db.close()

    return results
